File: backend/src/chatbots/chatbots.service.ts
import { Injectable } from '@nestjs/common';
import mysql from 'mysql2/promise';
import { RuntimeService } from '../runtime/runtime.service';
import { OcrDocumentsService } from '../ocr/ocr-documents.service';
import { DbConnectionsService } from '../db-connections/db-connections.service';
import { ChatbotGenericDto } from './dto/chatbot-generic.dto';
import { ChatbotOcrDto } from './dto/chatbot-ocr.dto';
import { ChatbotSqlDto } from './dto/chatbot-sql.dto';

@Injectable()
export class ChatbotsService {
  constructor(
    private readonly runtimeService: RuntimeService,
    private readonly ocrDocumentsService: OcrDocumentsService,
    private readonly dbConnectionsService: DbConnectionsService
  ) {}

  async generic(tenantId: string, dto: ChatbotGenericDto) {
    return this.runtimeService.execute(tenantId, {
      providerId: dto.providerId,
      model: dto.model,
      requestId: dto.requestId,
      payload: {
        messages: dto.messages
      }
    });
  }

  async ocr(tenantId: string, dto: ChatbotOcrDto) {
    const doc = await this.ocrDocumentsService.getById(tenantId, dto.documentId);
    if (!doc.enabled) {
      throw new Error('OCR document disabled');
    }
    const content = await this.ocrDocumentsService.getDecryptedContent(doc);
    const trimmed = content.slice(0, 12000);

    const messages = [
      {
        role: 'system',
        content:
          'Eres un asistente experto en lectura de documentos. Responde solo con la informacion del documento.'
      },
      {
        role: 'user',
        content: `Documento OCR:\\n${trimmed}\\n\\nPregunta: ${dto.question}`
      }
    ];

    return this.runtimeService.execute(tenantId, {
      providerId: dto.providerId,
      model: dto.model,
      requestId: dto.requestId,
      payload: { messages }
    });
  }

  async sql(tenantId: string, dto: ChatbotSqlDto) {
    const connection = await this.dbConnectionsService.getById(tenantId, dto.connectionId);
    if (!connection.enabled) {
      throw new Error('DB connection disabled');
    }

    const config = await this.dbConnectionsService.getDecryptedConfig(connection);
    const schema = await this.fetchSchemaSummary(connection, config);

    const sqlPrompt = [
      {
        role: 'system',
        content:
          'Eres un asistente que genera una unica consulta SQL de solo lectura (MySQL). Devuelve solo SQL sin explicaciones.'
      },
      {
        role: 'user',
        content: `Esquema:\\n${schema}\\n\\nPregunta: ${dto.question}`
      }
    ];

    let sql = dto.sql;
    if (!sql) {
      const response = await this.runtimeService.execute(tenantId, {
        providerId: dto.providerId,
        model: dto.model,
        requestId: dto.requestId,
        payload: { messages: sqlPrompt }
      });
      const output = response.output as any;
      const content = output?.choices?.[0]?.message?.content || output?.choices?.[0]?.text || '';
      sql = this.extractSql(content);
    }

    if (!sql) {
      throw new Error('No SQL generated');
    }

    this.validateReadOnly(sql, connection.readOnly);

    const rows = await this.executeQuery(config, sql);
    return {
      requestId: dto.requestId,
      sql,
      rows
    };
  }

  private extractSql(content: string) {
    const fenced = content.match(/```sql\\s*([\\s\\S]*?)```/i);
    if (fenced?.[1]) {
      return fenced[1].trim();
    }
    return content.trim();
  }

  private validateReadOnly(sql: string, readOnly: boolean) {
    if (!readOnly) {
      return;
    }
    const normalized = sql.trim().toLowerCase();
    const allowed = ['select', 'with', 'show', 'describe', 'explain'];
    if (!allowed.some((prefix) => normalized.startsWith(prefix))) {
      throw new Error('Only read-only queries are allowed');
    }
  }

  private async executeQuery(config: Record<string, unknown>, sql: string) {
    const connection = await mysql.createConnection({
      host: String(config.host || ''),
      port: Number(config.port || 3306),
      user: String(config.user || ''),
      password: String(config.password || ''),
      database: String(config.database || ''),
      ssl: config.ssl ? (config.ssl as any) : undefined
    });

    try {
      const [rows] = await connection.query(sql);
      return rows;
    } finally {
      await connection.end();
    }
  }

  private async fetchSchemaSummary(connection: { allowedTables: string[] }, config: Record<string, unknown>) {
    const allowed = connection.allowedTables ?? [];
    const db = String(config.database || '');
    if (!db) {
      return 'Database not specified';
    }

    const conn = await mysql.createConnection({
      host: String(config.host || ''),
      port: Number(config.port || 3306),
      user: String(config.user || ''),
      password: String(config.password || ''),
      database: db,
      ssl: config.ssl ? (config.ssl as any) : undefined
    });

    try {
      const [rows] = await conn.query(
        `SELECT TABLE_NAME as tableName, COLUMN_NAME as columnName, DATA_TYPE as dataType\n         FROM information_schema.COLUMNS\n         WHERE TABLE_SCHEMA = ?\n         ${allowed.length ? `AND TABLE_NAME IN (${allowed.map(() => '?').join(',')})` : ''}\n         ORDER BY TABLE_NAME, ORDINAL_POSITION\n         LIMIT 300`,
        allowed.length ? [db, ...allowed] : [db]
      );
      const lines: string[] = [];
      const grouped: Record<string, string[]> = {};
      (rows as any[]).forEach((row) => {
        const table = row.tableName;
        if (!grouped[table]) {
          grouped[table] = [];
        }
        grouped[table].push(`${row.columnName}:${row.dataType}`);
      });
      Object.entries(grouped).forEach(([table, cols]) => {
        lines.push(`${table}(${cols.join(', ')})`);
      });
      return lines.join('\\n');
    } finally {
      await conn.end();
    }
  }
}

File: backend-java/src/main/java/com/neria/manager/chatbots/ChatbotsService.java
package com.neria.manager.chatbots;

import com.neria.manager.dbconnections.DbConnectionsService;
import com.neria.manager.ocr.OcrDocumentsService;
import com.neria.manager.runtime.ExecuteRequest;
import com.neria.manager.runtime.RuntimeService;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class ChatbotsService {
  private final RuntimeService runtimeService;
  private final OcrDocumentsService ocrDocumentsService;
  private final DbConnectionsService dbConnectionsService;

  public ChatbotsService(
      RuntimeService runtimeService,
      OcrDocumentsService ocrDocumentsService,
      DbConnectionsService dbConnectionsService) {
    this.runtimeService = runtimeService;
    this.ocrDocumentsService = ocrDocumentsService;
    this.dbConnectionsService = dbConnectionsService;
  }

  public Object generic(String tenantId, ChatbotGenericRequest dto) {
    ExecuteRequest request = new ExecuteRequest();
    request.providerId = dto.providerId;
    request.model = dto.model;
    request.requestId = dto.requestId;
    request.payload = Map.of("messages", dto.messages);
    return runtimeService.execute(tenantId, request);
  }

  public Object ocr(String tenantId, ChatbotOcrRequest dto) {
    var doc = ocrDocumentsService.getById(tenantId, dto.documentId);
    if (!doc.isEnabled()) {
      throw new IllegalStateException("OCR document disabled");
    }
    String content = ocrDocumentsService.getDecryptedContent(doc);
    String trimmed = content.length() > 12000 ? content.substring(0, 12000) : content;

    List<Map<String, String>> messages =
        List.of(
            Map.of(
                "role",
                "system",
                "content",
                "Eres un asistente experto en lectura de documentos. Responde solo con la informacion del documento."),
            Map.of("role", "user", "content", "Documento OCR:\n" + trimmed + "\n\nPregunta: " + dto.question));

    ExecuteRequest request = new ExecuteRequest();
    request.providerId = dto.providerId;
    request.model = dto.model;
    request.requestId = dto.requestId;
    request.payload = Map.of("messages", messages);
    return runtimeService.execute(tenantId, request);
  }

  public Object sql(String tenantId, ChatbotSqlRequest dto) {
    var connection = dbConnectionsService.getById(tenantId, dto.connectionId);
    if (!connection.isEnabled()) {
      throw new IllegalStateException("DB connection disabled");
    }

    Map<String, Object> config = dbConnectionsService.getDecryptedConfig(connection);
    String schema = fetchSchemaSummary(dbConnectionsService.getAllowedTables(connection), config);

    List<Map<String, String>> prompt =
        List.of(
            Map.of(
                "role",
                "system",
                "content",
                "Eres un asistente que genera una unica consulta SQL de solo lectura (MySQL). Devuelve solo SQL sin explicaciones."),
            Map.of("role", "user", "content", "Esquema:\n" + schema + "\n\nPregunta: " + dto.question));

    String sql = dto.sql;
    if (sql == null || sql.isBlank()) {
      ExecuteRequest request = new ExecuteRequest();
      request.providerId = dto.providerId;
      request.model = dto.model;
      request.requestId = dto.requestId;
      request.payload = Map.of("messages", prompt);
      Object response = runtimeService.execute(tenantId, request);
      sql = extractSql(response);
    }

    if (sql == null || sql.isBlank()) {
      throw new IllegalStateException("No SQL generated");
    }

    validateReadOnly(sql, connection.isReadOnly());
    List<Map<String, Object>> rows = executeQuery(config, sql);
    return Map.of("requestId", dto.requestId, "sql", sql, "rows", rows);
  }

  private String extractSql(Object response) {
    if (!(response instanceof Map)) {
      return null;
    }
    Map<String, Object> map = (Map<String, Object>) response;
    Object output = map.get("output");
    if (!(output instanceof Map)) {
      return null;
    }
    Map<String, Object> outputMap = (Map<String, Object>) output;
    Object choicesObj = outputMap.get("choices");
    if (choicesObj instanceof List<?> choices && !choices.isEmpty()) {
      Object first = choices.get(0);
      if (first instanceof Map<?, ?> firstMap) {
        Object message = firstMap.get("message");
        if (message instanceof Map<?, ?> messageMap) {
          Object content = messageMap.get("content");
          if (content != null) {
            return stripSqlFence(String.valueOf(content));
          }
        }
        Object text = firstMap.get("text");
        if (text != null) {
          return stripSqlFence(String.valueOf(text));
        }
      }
    }
    Object responseField = outputMap.get("response");
    return responseField != null ? stripSqlFence(String.valueOf(responseField)) : null;
  }

  private String stripSqlFence(String content) {
    String trimmed = content.trim();
    if (trimmed.startsWith("```")) {
      int start = trimmed.indexOf("\n");
      int end = trimmed.lastIndexOf("```");
      if (start > 0 && end > start) {
        return trimmed.substring(start + 1, end).trim();
      }
    }
    return trimmed;
  }

  private void validateReadOnly(String sql, boolean readOnly) {
    if (!readOnly) {
      return;
    }
    String normalized = sql.trim().toLowerCase();
    List<String> allowed = List.of("select", "with", "show", "describe", "explain");
    boolean ok = allowed.stream().anyMatch(normalized::startsWith);
    if (!ok) {
      throw new IllegalStateException("Only read-only queries are allowed");
    }
  }

  private List<Map<String, Object>> executeQuery(Map<String, Object> config, String sql) {
    String host = String.valueOf(config.getOrDefault("host", ""));
    int port = Integer.parseInt(String.valueOf(config.getOrDefault("port", "3306")));
    String user = String.valueOf(config.getOrDefault("user", ""));
    String password = String.valueOf(config.getOrDefault("password", ""));
    String database = String.valueOf(config.getOrDefault("database", ""));
    String jdbc = "jdbc:mysql://" + host + ":" + port + "/" + database;

    try (Connection connection = DriverManager.getConnection(jdbc, user, password);
        Statement statement = connection.createStatement()) {
      ResultSet rs = statement.executeQuery(sql);
      List<Map<String, Object>> rows = new ArrayList<>();
      int columnCount = rs.getMetaData().getColumnCount();
      while (rs.next()) {
        java.util.Map<String, Object> row = new java.util.HashMap<>();
        for (int i = 1; i <= columnCount; i++) {
          row.put(rs.getMetaData().getColumnLabel(i), rs.getObject(i));
        }
        rows.add(row);
      }
      return rows;
    } catch (Exception ex) {
      throw new IllegalStateException("SQL execution failed: " + ex.getMessage(), ex);
    }
  }

  private String fetchSchemaSummary(List<String> allowedTables, Map<String, Object> config) {
    String host = String.valueOf(config.getOrDefault("host", ""));
    int port = Integer.parseInt(String.valueOf(config.getOrDefault("port", "3306")));
    String user = String.valueOf(config.getOrDefault("user", ""));
    String password = String.valueOf(config.getOrDefault("password", ""));
    String database = String.valueOf(config.getOrDefault("database", ""));
    if (database == null || database.isBlank()) {
      return "Database not specified";
    }
    String jdbc = "jdbc:mysql://" + host + ":" + port + "/" + database;
    StringBuilder sql = new StringBuilder();
    sql.append(
        "SELECT TABLE_NAME as tableName, COLUMN_NAME as columnName, DATA_TYPE as dataType " +
        "FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = ?");
    if (allowedTables != null && !allowedTables.isEmpty()) {
      sql.append(" AND TABLE_NAME IN (");
      sql.append(String.join(",", java.util.Collections.nCopies(allowedTables.size(), "?")));
      sql.append(")");
    }
    sql.append(" ORDER BY TABLE_NAME, ORDINAL_POSITION LIMIT 300");

    try (Connection connection = DriverManager.getConnection(jdbc, user, password);
        PreparedStatement statement = connection.prepareStatement(sql.toString())) {
      int index = 1;
      statement.setString(index++, database);
      if (allowedTables != null) {
        for (String table : allowedTables) {
          statement.setString(index++, table);
        }
      }
      ResultSet rs = statement.executeQuery();
      java.util.Map<String, List<String>> grouped = new java.util.HashMap<>();
      while (rs.next()) {
        String table = rs.getString("tableName");
        grouped.computeIfAbsent(table, ignored -> new ArrayList<>())
            .add(rs.getString("columnName") + ":" + rs.getString("dataType"));
      }
      StringBuilder output = new StringBuilder();
      grouped.forEach((table, cols) -> {
        if (!output.isEmpty()) {
          output.append("\n");
        }
        output.append(table).append("(").append(String.join(", ", cols)).append(")");
      });
      return output.toString();
    } catch (Exception ex) {
      throw new IllegalStateException("Schema fetch failed: " + ex.getMessage(), ex);
    }
  }

  public static class ChatbotGenericRequest {
    public String providerId;
    public String model;
    public String requestId;
    public List<Map<String, String>> messages;
  }

  public static class ChatbotOcrRequest {
    public String providerId;
    public String model;
    public String requestId;
    public String documentId;
    public String question;
  }

  public static class ChatbotSqlRequest {
    public String providerId;
    public String model;
    public String requestId;
    public String connectionId;
    public String question;
    public String sql;
  }
}
